src/ekat/logging/ekat_logger.hpp

```diff
@@ -0,0 +1,158 @@
+#ifndef EKAT_LOGGER_HPP
+#define EKAT_LOGGER_HPP
+
+#include "spdlog/spdlog.h"
+#include "ekat/logging/ekat_log_file.hpp"
+#include "ekat/logging/ekat_log_mpi.hpp"
+#include "spdlog/sinks/stdout_color_sinks.h"
+#include "spdlog/sinks/null_sink.h"
+
+namespace ekat {
+namespace logger {
+
+// expose the default logger; this is the easiest one to use
+namespace Log = spdlog;
+
+/* A Logger customized with policies for console and file output.
+
+  The default Logger<> sets up colored output to the console only (no log files), for all mpi ranks.
+
+  Logs have a priority logging level.  Messages with priorities below this level will
+  be ignored.
+
+  Logging levels (increasing priority):
+  - off
+  - trace
+  - debug
+  - info
+  - warn
+  - err
+  - critical
+
+  The default priority for the spdlog library is "info."  For EKAT, we've set the
+  default to "debug."  Both are easy to change.
+
+  The LogMpiPolicy can restrict console output to only rank 0, and it will append
+  rank ids to log files and log names.
+
+  The LogFilePolicy determines whether or not the log will also write output to a file.
+  Files and console can have different logging levels.  For example, the
+  log level for the LogBasicFile policy could be "debug," while the level of the console
+  logger could be set to "info".  In this case, no debug messages will show up on the
+  console, but they will be in the log files.
+
+  In principle, separate modules, classes, etc. could have their own logger.
+  These loggers can share output files, too; see the spdlog wiki
+(https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-multiple-file-loggers-with-the-same-output-file) for an example.
+
+  The ekat::Logger class is a subclass of spdlog::logger class, to simplify the many
+  features of that library to the ones EKAT is most likely to use, while retaining all
+  of its functionality.
+*/
+template <typename LogFilePolicy = LogNoFile,    // see ekat_log_file.hpp
+          typename LogMpiPolicy = LogIgnoreRank> // see ekat_log_mpi.hpp
+class Logger : public spdlog::logger {
+  public:
+    Logger() = delete;
+
+    Logger(const std::string& log_name, const std::string& level_str="debug") :
+     spdlog::logger(LogMpiPolicy::name_append_rank(log_name))
+  {
+
+    const auto clevel = Log::level::from_str(level_str);
+    auto csink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
+    csink->set_level(clevel);
+
+    std::string fname = LogMpiPolicy::name_append_rank(log_name) + "_logfile.txt";
+    auto fsink = LogFilePolicy::get_file_sink(fname);
+
+    this->sinks().push_back(csink);
+    this->sinks().push_back(fsink);
+
+    LogMpiPolicy::update_sinks(this->sinks());
+
+    this->set_level(clevel);
```

Comment:
Question (on spdlog): I'm a bit confused by the fact that we can set a log level on both the logger and the sink(s). I get that different sinks may have different log levels, so that's fine, but why also a level on the Logger? Is the idea that the Logger defines some sort of "general" level, and then each sink can have its own "additional" way of sift logging?

Reply:
This confused me too, at first.  I had to write some extra tests to verify I understood the behavior.   It's an extra layer of flexibility that has some nice consequences.  In pseudocode, the logic flows like this:

```
def log_message(msg):
  if msg.priority >= log.priority:
    sink_msg(msg)
  else:
    return

def sink_msg(msg):
  if msg.priority >= sink.priority
    write_msg(msg)
  else:
   return
```

In hindsight, that pseudocode is probably obvious.  But here's how it becomes helpful.  Let's say Hommexx and Haero each have a logger, but they share the same sinks (console and file).   Since Hommexx is a fancy SC20 code, we have confidence in it and we don't  need to see all of its debug and trace output, so we set the hommexx logger level to `warn`.  Haero is just getting started though and it's largely unproven, so we set the Haero logger level to `debug`.   This means that the log files will only show warning or higher priority messages from Hommexx, but they'll show debug output from haero, all in the same sinks.

 If the file sink were the same level as the logger, then either all the haero debug output would be surpassed by the higher logging level of hommexx, or we'd see all of hommexx's debug output.

Btw, the test code I wrote to verify this is in `logger_tests.cpp`, lines 53--55:

```
      // the file level is trace, but the log level is debug (debug > trace); trace messages will be skipped.
      mylog.trace("this message won't show up anywhere.");
      REQUIRE( !mylog.should_log(Log::level::trace) );
```

Reply:
This makes sense. Thanks for the detailed explanation!